modules/liana.nix

```diff
@@ -0,0 +1,102 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+let
+  options.services.lianad = {
+    enable = mkEnableOption "lianad bitcoin wallet";
+    daemon = mkOption {
+      type = types.bool;
+      default = false;
+      description = mdDoc "Whether to run the process as a UNIX daemon (double fork magic).";
+    };
+    data_dir = mkOption {
+      type = types.path;
+      default = "/var/lib/lianad";
+      description = mdDoc "Path to the folder where we should store the application data.";
+    };
+    main_descriptor = mkOption {
+      type = types.str;
+      default = "wsh(or_d(pk([0dd8c6f0/48'/1'/0'/2']tpubDFMbZ7U5k5hEfsttnZTKMmwrGMHnqUGxhShsvBjHimXBpmAp5KmxpyGsLx2toCaQgYq5TipBLhTUtA2pRSB9b14m5KwSohTDoCHkk1EnqtZ/<0;1>/*),and_v(v:pkh([d4ab66f1/48'/1'/0'/2']tpubDEXYN145WM4rVKtcWpySBYiVQ229pmrnyAGJT14BBh2QJr7ABJswchDicZfFaauLyXhDad1nCoCZQEwAW87JPotP93ykC9WJvoASnBjYBxW/<0;1>/*),older(65535))))#7nvn6ssc";
+      description = mdDoc "The wallet descriptor.";
+    };
+    network = mkOption {
+      type = types.str;
+      default = "bitcoin";
+      description = mdDoc "bitcoin, testnet, signet, or regtest";
+    };
+    bitcoind_addr = mkOption {
+      type = types.str;
+      default = "127.0.0.1";
+      description = mdDoc "bitcoind address.";
+    };
+    bitcoind_port = mkOption {
+      type = types.port;
+      default = 8332;
+      description = mdDoc "bitcoind port.";
+    };
+  };
+
+  cfg = config.services.lianad;
+  nbLib = config.nix-bitcoin.lib;
+  secretsDir = config.nix-bitcoin.secretsDir;
+  bitcoind = config.services.bitcoind;
+in {
+  inherit options;
+
+  config = mkIf cfg.enable {
+    services.bitcoind = {
+      enable = true;
+      listenWhitelisted = true;
+    };
+
+    systemd.tmpfiles.rules = [
+      "d '${cfg.dataDir}' 0770 ${cfg.user} ${cfg.group} - -"
+    ];
+
+    systemd.services.lianad = {
+      wantedBy = [ "multi-user.target" ];
+      requires = [ "bitcoind.service" ];
+      after = [ "bitcoind.service" "nix-bitcoin-secrets.target" ];
+      preStart = ''
+        cat << EOF > lianad_config.toml
+# these should come from options.services.lianad
+daemon = false
+data_dir = "/var/lib/lianad"
+log_level = "debug"
+main_descriptor = "wsh(or_d(pk([0dd8c6f0/48'/1'/0'/2']tpubDFMbZ7U5k5hEfsttnZTKMmwrGMHnqUGxhShsvBjHimXBpmAp5KmxpyGsLx2toCaQgYq5TipBLhTUtA2pRSB9b14m5KwSohTDoCHkk1EnqtZ/<0;1>/*),and_v(v:pkh([d4ab66f1/48'/1'/0'/2']tpubDEXYN145WM4rVKtcWpySBYiVQ229pmrnyAGJT14BBh2QJr7ABJswchDicZfFaauLyXhDad1nCoCZQEwAW87JPotP93ykC9WJvoASnBjYBxW/<0;1>/*),older(65535))))#7nvn6ssc"
+
+# these should come from options.services.lianad
+[bitcoin_config]
+network = "signet"
+poll_interval_secs = 30
+
+# these should come from options.services.bitcoind
+[bitcoind_config]
+addr = "127.0.0.1:38332"
+auth = "username:password"
+
+EOF
+        '';
```

Comment:
This could be a `builtins.fromTOML` follwed by a `toString`, e.g.:

```nix
let
  tomlFile =./example.toml;
  tomlData = builtins.readFile tomlFile;
  tomlObj = builtins.fromTOML tomlData;
  tomlString = toString tomlObj;
in
  tomlString
```

Then you can interpolate the hardcoded nastiness with `options.services.lianad.*`

Reply:
Refs:

- https://nix.dev/manual/nix/2.22/language/builtins.html#builtins-fromTOML
- https://nix.dev/manual/nix/2.22/language/builtins.html#builtins-readFile
- https://nix.dev/manual/nix/2.22/language/builtins.html#builtins-toString

Reply:
will this allow me to dynamically generate the TOML file and write it into the output's datadir?

it seems like a convenient tool for reading configs that already exist on disk, but our use case is the opposite direction:

we need to generate a TOML file with values coming from `options.services.lianad.*`

Reply:
You can generate a static toml file from a set using ``pkgs.formats.toml``. E.g:

```nix
mySet = { foo = "bar""; };
myTomlFile = (pkgs.formats.toml {}).generate mySet;
```

But this of course cannot contain secrets such as bitcoind credentials that are not going into the nix store. So then the `preStart` rule could execute a command that reads the generated TOML, and appends the additional secrets, writing it to the dataDir for liana.

The lnd module actually already does this:

```nix
      preStart = ''
        install -m600 ${configFile} '${cfg.dataDir}/lnd.conf'
        {
          echo "bitcoind.rpcpass=$(cat ${secretsDir}/bitcoin-rpcpassword-${rpcUser})"
          ${optionalString (cfg.getPublicAddressCmd != "") ''
            echo "externalip=$(${cfg.getPublicAddressCmd})"
          ''}
        } >> '${cfg.dataDir}/lnd.conf'
```